x/feerefunder/keeper/keeper_test.go

```diff
@@ -156,18 +162,80 @@ func TestKeeperLockFees(t *testing.T) {
-	bankKeeper.EXPECT().SendCoinsFromAccountToModule(ctx, payer, types.ModuleName, validFee.Total()).Return(fmt.Errorf("bank error"))
-	err = k.LockFees(ctx, payer, packet, validFee)
+	bankKeeper.EXPECT().SendCoinsFromAccountToModule(ctx, payerInfo.Sender, types.ModuleName, validFee.Total()).Return(fmt.Errorf("bank error"))
+	err = k.LockFees(ctx, payerInfo, packet, validFee)
 	require.ErrorContains(t, err, "bank error")
 
 	channelKeeper.EXPECT().GetChannel(ctx, packet.PortId, packet.ChannelId).Return(channeltypes.Channel{}, true)
-	bankKeeper.EXPECT().SendCoinsFromAccountToModule(ctx, payer, types.ModuleName, validFee.Total()).Return(nil)
-	err = k.LockFees(ctx, payer, packet, validFee)
+	bankKeeper.EXPECT().SendCoinsFromAccountToModule(ctx, payerInfo.Sender, types.ModuleName, validFee.Total()).Return(nil)
+	err = k.LockFees(ctx, payerInfo, packet, validFee)
+	require.NoError(t, err)
+	require.Equal(t, sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeLockFees,
+			sdk.NewAttribute(types.AttributeKeyPayer, payerInfo.Sender.String()),
+			sdk.NewAttribute(types.AttributeKeyPortID, packet.PortId),
+			sdk.NewAttribute(types.AttributeKeyChannelID, packet.ChannelId),
+			sdk.NewAttribute(types.AttributeKeySequence, strconv.FormatUint(packet.Sequence, 10)),
+		),
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+		),
+	}, ctx.EventManager().Events())
+}
+
+func TestKeeperLockFeesAtFeePayeer(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	bankKeeper := mock_types.NewMockBankKeeper(ctrl)
+	channelKeeper := mock_types.NewMockChannelKeeper(ctrl)
+	feegrantKeeper := mock_types.NewMockFeeGrantKeeper(ctrl)
+	feeAllowance := mock_feegrant.NewMockFeeAllowanceI(ctrl)
+	k, ctx := testutil_keeper.FeeKeeper(t, channelKeeper, bankKeeper, feegrantKeeper)
+
+	payerInfo := types.PayerInfo{
+		Sender:   sdk.MustAccAddressFromBech32(testutil.TestOwnerAddress),
+		FeePayer: sdk.MustAccAddressFromBech32(TestAddress),
+	}
+
+	k.SetParams(ctx, types.Params{
+		MinFee: types.Fee{
+			RecvFee:    nil,
+			AckFee:     sdk.NewCoins(sdk.NewCoin("denom1", sdk.NewInt(100)), sdk.NewCoin("denom2", sdk.NewInt(100))),
+			TimeoutFee: sdk.NewCoins(sdk.NewCoin("denom1", sdk.NewInt(100)), sdk.NewCoin("denom2", sdk.NewInt(100))),
+		},
+	})
+
+	packet := types.PacketID{
+		ChannelId: "channel-0",
+		PortId:    "transfer",
+		Sequence:  111,
+	}
+	ackFee := sdk.NewCoins(sdk.NewCoin("denom1", sdk.NewInt(101)))
+	timeoutFee := sdk.NewCoins(sdk.NewCoin("denom1", sdk.NewInt(101)))
+	validFee := types.Fee{
+		RecvFee:    nil,
+		AckFee:     ackFee,
+		TimeoutFee: timeoutFee,
+	}
+
+	channelKeeper.EXPECT().GetChannel(ctx, packet.PortId, packet.ChannelId).Return(channeltypes.Channel{}, true)
+	feegrantKeeper.EXPECT().GetAllowance(ctx, payerInfo.FeePayer, payerInfo.Sender).Return(nil, fmt.Errorf("feegrant error"))
+	err := k.LockFees(ctx, payerInfo, packet, validFee)
+	require.ErrorContains(t, err, "feegrant error")
```

Comment:
how about adding one more case for when allowance.Accept() returns error? It actually is pretty real, cause it's when there are no more coins left to grant error